content/en/blog/_posts/2022-11-11-retroactive-default-storage-class.md

```diff
@@ -0,0 +1,131 @@
+
+
+   2. Remove the old default first and create new default StorageClass which results in having no default for a short period of time. If user creates a PersistentVolumeClaim with storageClassName set to `nil` (which means default StorageClass) the PVC will be in `Pending` state forever, then user has to fix this by deleting the PVC a creating it again once default StorageClass is available.
+
+Looking at the scenarios above it is clear that neither of them is optimal.
+
+### Resource ordering during cluster installation
+
+If cluster installation tool needs to create resources that require storage (e.g. image registry) it was difficult to get the ordering right because any Pods that require storage and rely on default StorageClass presence will fail to create if there is no default at that point yet.
+
+## What changed
+
+We've changed PersistentVolume (PV) controller in KCM, so it can assign default StorageClass to unbound PersistentVolumeClaims that have storageClassName set to `nil`. In order to allow the value change PersistentVolumeClaim admission in the API server also changed to allow changing the value from `nil` to actual StorageClass name.
+
+### `nil` vs `""` for storageClassName - does it matter?
+
+Before this feature was introduced those values were equal in terms of behavior and any PersistentVolumeClaim with `nil` or `""` storageClassName value would bind to existing PersistentVolume resource with storageClassName also set to `nil` or `""`.
+
+With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name. With these constraints in mind the feature changes the semantics of `nil` if default StorageClass is present to always mean "Give me a default" and `""` to mean "Give me PersistentVolume that also has `""` StorageClass name".
+
+In other words we need to distinguish two main cases, either default StorageClass is not present in which case there's no behavior change or where there is a default in which case PVCs with `nil` won't bind to PVs anymore and instead will be updated to have the StorageClass.
+
+The tables below show all these cases to better describe when PVC binds and when it's StorageClass gets updated.
+
+| Without default class        | PVC storageClassName == `""` | PVC storageClassName == `nil` |
+|------------------------------|------------------------------|-------------------------------|
+| PV storageClassName == `""`  | binds                        | binds                         |
+| PV without storageClassName  | binds                        | binds                         |
+
+
+| With default class          | PVC storageClassName == `""` | PVC storageClassName == `nil` |
```

Comment:
Can you add a title for each of these two tables so we know which has the new feature enabled and which does not?

Reply:
It might be better to use a custom HTML table where the first column uses `rowspan`, and have that in place of these two tables.

Alternatively, have the table in 2 sections, each section introduced by a cell spanning all columns.

Reply:
I've changed the table to HTML (first version you suggested). Not sure I got the idea right though.